### PR TITLE
planner: fixing wrong result after applying predicate push down for CTEs

### LINKED ISSUE #47881: Predicate conditions are incorrectly pushed down when CTE is referenced multiple times.
## Bug Report

Predicate conditions are incorrectly pushed down when CTE is referenced multiple times.
image

### 1. Minimal reproduce step (Required)
```
create table t (id int,name varchar(10));
insert into t values(1,'tt');
create table t1(id int,name varchar(10),name1 varchar(10),name2 varchar(10));
insert into t1 values(1,'tt','ttt','tttt'),(2,'dd','ddd','dddd');
create table t2(id int,name varchar(10),name1 varchar(10),name2 varchar(10),`date1` date);
insert into t2 values(1,'tt','ttt','tttt','2099-12-31'),(2,'dd','ddd','dddd','2099-12-31');
WITH bzzs
  AS
    (
           SELECT count(1) AS bzn
           FROM   t c
            ) ,
  tmp1
  AS
    (
              SELECT    t1.*
              FROM      t1
              LEFT JOIN bzzs
              ON        1=1
              WHERE     name IN ('tt')
              AND       bzn <> 1 ) ,
  tmp2
  AS
    (
             SELECT   tmp1.*,date('2099-12-31') AS endate
             FROM     tmp1) ,
  tmp3
  AS
    (
           SELECT *
           FROM   tmp2
           WHERE  endate > CURRENT_DATE
           UNION ALL
           SELECT '1'                AS id ,
                  'ss'  AS name ,
                  'sss' AS name1,
                  'ssss' AS name2,
                  date('2099-12-31') AS endate
           FROM   bzzs t1
           WHERE  bzn = 1 )
    SELECT    c2.id ,
              c3.id
    FROM      t2 db
    LEFT JOIN tmp3 c2
    ON        c2.id='1'
    LEFT JOIN tmp3 c3
    ON        c3.id='1';
```

### 2. What did you expect to see? (Required)
```
+------+------+
| id   | id   |
+------+------+
| 1    | 1    |
| 1    | 1    |
+------+------+
```

### 3. What did you see instead (Required)
```
+------+------+
| id   | id   |
+------+------+
| NULL | NULL |
| NULL | NULL |
+------+------+
```
### 4. What is your TiDB version? (Required)

v6.5.4

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve?

Issue Number: close #47881 

Problem Summary:

### What is changed and how it works?

The `outerMost` should be prop of the CTE itself, not the CTE reader's.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  >   

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Fix the issue that TiDB may push wrong predicates into the CTE.
```

### CODE REVIEW

**Review by time-and-fate — APPROVED** [review]
Please use a more meaningful pr title.

**Review by elsa0520 — APPROVED** [review]
LGTM

**Review by fixdb — APPROVED** [review]
LGTM

**Review by AilinKid — APPROVED** [review]

**Comment by AilinKid** [comment]
we can add some more comment to depict it
```
use a tree chart like
cte_tree1  <---+ (consuming)
cte_tree2  ----+   <--+  (consuming)
cte_tree3 ------------+
this way, the outermost cte is only cte_tree3
```

File: pkg/planner/core/recheck_cte.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+	if cteReader, ok := p.(*LogicalCTE); ok {
+		cte := cteReader.cte
+		if !isRootTree {
+			inDegree[cte.IDForStorage]++
+		}
+		if checked.Has(cte.IDForStorage) {
+			return
+		}
+		ctes[cte.IDForStorage] = cte
+		checked.Insert(cte.IDForStorage)
```
Comment (time-and-fate) [inline pkg/planner/core/recheck_cte.go:48]: I think we don't need this `checked *intset.FastIntSet` if we already have `ctes map[int]*CTEClass`

File: pkg/planner/core/recheck_cte.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
+// Copyright 2016 PingCAP, Inc.
```
Comment (time-and-fate) [inline pkg/planner/core/recheck_cte.go:1]:
```suggestion
// Copyright 2023 PingCAP, Inc.
```

File: pkg/planner/core/recheck_cte.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import "github.com/pingcap/tidb/pkg/util/intset"
+
+// RecheckCTE fills the IsOuterMostCTE field for CTEs.
+// It's a temp solution to before we fully use the Sequence to optimize the CTEs.
+// This func will find the dependent relation of each CTEs.
```
Comment (time-and-fate) [inline pkg/planner/core/recheck_cte.go:21]: I think this is a bit misleading, we are not looking for "the dependent relation of each CTE" here, we are just checking if it is at the outermost level for each CTE.

**Review by time-and-fate — APPROVED** [review]

**Pull request merged**